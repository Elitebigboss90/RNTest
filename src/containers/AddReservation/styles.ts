import { StyleSheet } from "react-native";

const styles = StyleSheet.create({
    container: {
        flex: 1,
    },
    headerIcon: {
        height: 24,
        width: 24,
        transform: [{ rotate: '90deg' }]
    },
    textColor: {
        color: 'black'
    },
    textInput: {
        height: 50,
        margin: 8,
        borderBottomWidth: 1,
        borderBottomColor: 'grey'
    }
});

export default styles;
