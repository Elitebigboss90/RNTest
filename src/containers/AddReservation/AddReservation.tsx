import React, { Component } from "react";
import {
    View,
    Text,
    TextInput,
    TouchableOpacity,
    Alert,
    DatePickerIOS,
    DatePickerAndroid,
    Platform
} from "react-native";
import PropTypes from "prop-types";

import { Mutation } from 'react-apollo';
import gql from 'graphql-tag';
import { NavigationScreenProp } from "react-navigation";

import { Header } from "../../common/Header/Header";
import { Button } from "../../common/Button/Button";
import { nextIcon } from "../../static/imports/icons";
import styles from "./styles";
import console from "console";
//import console = require("console");



interface IProps {
    navigation: NavigationScreenProp<any, any>;
}

interface IState {
    name: string,
    hotelName: string,
    arrivalDate: Date,
    departureDate: Date,
    arrivalDateISO: string,
    departureDateISO: string,
    [key: string]: any
}

const Add_Reservation = gql`mutation PostMutation($name: String!, $hotelName: String!,$arrivalDate: String!, $departureDate: String!){
    createReservation(data:{name:$name,hotelName:$hotelName,arrivalDate:$arrivalDate,departureDate:$departureDate}){
      name
      hotelName
      arrivalDate
      departureDate
    }
  }
`

class AddReservationScreen extends Component<IProps, IState>{

    constructor(props: IProps) {

        super(props);
        this.state = {
            name: "",
            hotelName: "",
            arrivalDate: new Date(),
            departureDate: new Date(),
            arrivalDateISO: "",
            departureDateISO: "",

        }

        this._handleMultiInput = this._handleMultiInput.bind(this);
        this._onSubmit = this._onSubmit.bind(this);
    }

    public static propTypes: any = {
        navigation: PropTypes.shape({}).isRequired,
    };

    public handleNavigation = () => (e: Event) => {
        const { navigation } = this.props;
        navigation.goBack()
    }

    public _handleMultiInput(stateName: string) {
        var id: string = stateName;
        return (text: string) => {
            var value: string = text;
            this.setState({ [id]: value })
        }
    }

    public _handleDateInput(stateName: string) {
        var name: string = stateName;
        return (selectedDate: Date) => {
            var value: Date = selectedDate;
            this.setState({ [name]: value })
        }

    }

    public getFormattedDate(date: Date) {
        var year = date.getFullYear();

        var month = (1 + date.getMonth()).toString();
        month = month.length > 1 ? month : '0' + month;

        var day = date.getDate().toString();
        day = day.length > 1 ? day : '0' + day;

        return month + '/' + day + '/' + year;
    }

    public _handleButton = () => {

    }

    _onCompleted = (data: any) => {
        const { navigation } = this.props;
        navigation.goBack()
    }

    public datePickerAndroid = (stateName: string) => async e => {
        try {
            const { action, year, month, day } = await DatePickerAndroid.open({
                // Use `new Date()` for current date.
                // May 25 2020. Month 0 is January.
                date: new Date()
            });
            console.log("android", action)
            Alert.alert("here", date)
            if (action !== DatePickerAndroid.dismissedAction) {
                // Selected year, month (0-11), day
                let dateString = (parseInt(month) + 1) + '/' + day + '/' + year;
                this.setState({ [stateName]: dateString })
            }
        } catch ({ code, message }) {
            console.warn('Cannot open date picker', message);
        }
    }


    public datePickerController() {
        const { arrivalDate, departureDate } = this.state;
        return (
            <>
                {Platform.OS === 'ios' ? <><Text>Arrival Date</Text><DatePickerIOS
                    date={arrivalDate}
                    onDateChange={this._handleDateInput('arrivalDate')}
                    mode={'date'}
                /></> : <TouchableOpacity onPress={this.datePickerAndroid('arrivalDateISO')}><Text>Add ArrivalDate</Text></TouchableOpacity>}
                {Platform.OS === 'ios' ? <><Text>Departure Date</Text><DatePickerIOS
                    date={departureDate}
                    mode={'date'}
                    minimumDate={arrivalDate}
                    onDateChange={this._handleDateInput('departureDate')}
                /></> : <TouchableOpacity onPress={this.datePickerAndroid('departureDateISO')}><Text>Add Departure Date</Text></TouchableOpacity>}
            </>
        )
    }
    public _onSubmit = (mutation: any) => (e: Event) => {

        const { arrivalDate, departureDate, name, hotelName } = this.state;
        if (Platform.OS === 'ios') {
            let arrivalDateISO = this.getFormattedDate(arrivalDate);
            let departureDateISO = this.getFormattedDate(departureDate);

            this.setState({ arrivalDateISO, departureDateISO }, () => {
                if (name && hotelName && arrivalDateISO && departureDateISO)
                    mutation();
                else {
                    Alert.alert("Please Fill Out Blank(s)")
                }
            })
        } else {
            if (name && hotelName)
                mutation();
            else {
                Alert.alert("Please Fill Out Blank(s)")
            }
        }

    }

    public onError(error: any) {
        /**
         * TODO: ERROR HANDLER FOR GRAPHQL REQUEST FAILD
         */
    }

    public render() {
        const { name, hotelName, arrivalDateISO, departureDateISO } = this.state;

        return (
            <Mutation mutation={Add_Reservation} onError={this.onError} onCompleted={this._onCompleted} variables={{ name, hotelName, arrivalDate: arrivalDateISO, departureDate: departureDateISO }}>{PostMutation =>
                <View style={styles.container}>
                    <Header>
                        <Button icon={nextIcon} onPress={this.handleNavigation} style={styles.headerIcon} />
                        <View>
                            <Text>Add New Reservation</Text>
                        </View>
                        <View />
                    </Header >
                    <View>
                        <TextInput
                            style={styles.textInput}
                            onChangeText={this._handleMultiInput('name')}
                            value={name}
                        />
                        <TextInput
                            style={styles.textInput}
                            onChangeText={this._handleMultiInput('hotelName')}
                            value={hotelName}
                        />
                        {this.datePickerController()}
                        {/**
                         *TODO:Add helper text to show the dates
                         */}
                        <TouchableOpacity onPress={this._onSubmit(PostMutation)}>
                            <Text>Submit</Text>
                        </TouchableOpacity>
                    </View>
                </View>}
            </Mutation>
        )
    }

}

export default AddReservationScreen;