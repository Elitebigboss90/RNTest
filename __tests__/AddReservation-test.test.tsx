import React from 'react';
import AddReservation from '../src/containers/AddReservation/AddReservation';
import { shallow, ShallowWrapper } from "enzyme";
import { View } from "react-native";

import { MockedProvider } from 'react-apollo/test-utils';

import renderer from 'react-test-renderer';
import console = require('console');

const createTestProps = (props: Object) => ({
    navigation: {
        state: { params: {} },
        dispatch: jest.fn(),
        goBack: jest.fn(),
        dismiss: jest.fn(),
        navigate: jest.fn(),
        openDrawer: jest.fn(),
        closeDrawer: jest.fn(),
        toggleDrawer: jest.fn(),
        getParam: jest.fn(),
        setParams: jest.fn(),
        addListener: jest.fn(),
        push: jest.fn(),
        replace: jest.fn(),
        pop: jest.fn(),
        popToTop: jest.fn(),
        isFocused: jest.fn()
    },
    ...props
});




describe("AddReservationScreen", () => {
    describe("rendering", () => {
        let wrapper: ShallowWrapper;
        let props: any;   // use type "any" to opt-out of type-checking
        let navigation: any;
        beforeEach(() => {
            props = createTestProps({});
            navigation = { navigate: jest.fn() };
            wrapper = shallow(<AddReservation navigation={navigation} {...props} />);

        });

        it('should render the landing page without crashing', () => {
            const rendered = renderer.create(<MockedProvider mocks={[]}><AddReservation navigation={navigation} {...props} /></MockedProvider>).toJSON();
            expect(rendered).toBeTruthy();
        })

        it('should render the child element', () => {
            //const wrapper = shallow(<AddReservation navigation={navigation} {...props} />);
            const rendered = renderer.create(<MockedProvider mocks={[]}><AddReservation navigation={navigation} {...props} /></MockedProvider>);
            const renderedInstance = rendered.root;
            expect(rendered).toMatchSnapshot();
            expect(renderedInstance.findByProps({ testID: '#name' })).toBeTruthy();
        })

        it("should check the states if mounted right", () => {
            expect(wrapper.state('arrivalDateISO')).toBe("");
        });

        it("should render a mounted Mutation", () => {
            expect(wrapper.find('Mutation')).toHaveLength(1);
        });

        /*it("should check the navigation if it's goes back", () => {
            const spy = jest.spyOn(navigation, 'navigate')
            const backButton = wrapper.find('Button').at(0);
            backButton.props().onClick;
            expect(spy).toBeCalledWith('Reservation');
            //expect(props.navigation.navigate).toHaveBeenCalledWith('Reservation');   // SUCCESS
        });*/

        it('should check the navigation if its goes back', () => {
            const component = renderer.create(<MockedProvider mocks={[]}><AddReservation navigation={navigation} {...props} /></MockedProvider>);
            const backButton = component.root.findByProps({ testID: '#back' })
            backButton.props.onPress();
            expect(component).toMatchSnapshot();
            expect(props.navigation.goBack).toHaveBeenCalled();
            //@ts-ignore
            //component._handleNameInput('ethan');
            //@ts-ignore
            //expect(component.state.name).toEqual('ethan')
        });

        it('should check the nameInput if its update state correctly', () => {
            const component = shallow(<AddReservation navigation={navigation} {...props} />);
            const componentInstance = component.instance();
            expect(component).toMatchSnapshot();
            //const spy = jest.spyOn(AddReservation.prototype, '_handelName');
            //@ts-ignore
            componentInstance._handleNameInput('ethan');
            //@ts-ignore
            expect(componentInstance.state.name).toEqual('ethan');
        });
    });
});
