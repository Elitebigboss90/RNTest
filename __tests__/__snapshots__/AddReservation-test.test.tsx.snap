// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddReservationScreen rendering should check the nameInput if its update state correctly 1`] = `ShallowWrapper {}`;

exports[`AddReservationScreen rendering should check the navigation if its goes back 1`] = `
<View
  style={
    Object {
      "flex": 1,
    }
  }
>
  <RCTSafeAreaView
    emulateUnlessSupported={true}
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "white",
          "borderBottomColor": "grey",
          "borderBottomWidth": 1,
          "flexDirection": "row",
          "justifyContent": "space-between",
          "minHeight": 64,
          "paddingHorizontal": 8,
        }
      }
    >
      <View
        accessible={true}
        isTVSelectable={true}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "height": 24,
            "opacity": 1,
            "transform": Array [
              Object {
                "rotate": "90deg",
              },
            ],
            "width": 24,
          }
        }
        testID="#back"
      >
        <Image
          source={
            Object {
              "testUri": "../../../src/static/icons/iconBlackChevronRight.png",
            }
          }
          style={
            Object {
              "height": 24,
              "transform": Array [
                Object {
                  "rotate": "90deg",
                },
              ],
              "width": 24,
            }
          }
        />
      </View>
      <View>
        <Text>
          Add New Reservation
        </Text>
      </View>
      <View />
    </View>
  </RCTSafeAreaView>
  <View>
    <TextInput
      allowFontScaling={true}
      onChangeText={[Function]}
      style={
        Object {
          "borderBottomColor": "grey",
          "borderBottomWidth": 1,
          "height": 50,
          "margin": 8,
        }
      }
      testID="#name"
      underlineColorAndroid="transparent"
      value=""
    />
    <TextInput
      allowFontScaling={true}
      onChangeText={[Function]}
      style={
        Object {
          "borderBottomColor": "grey",
          "borderBottomWidth": 1,
          "height": 50,
          "margin": 8,
        }
      }
      testID="#hotelname"
      underlineColorAndroid="transparent"
      value=""
    />
    <Text>
      Arrival Date
    </Text>
    <View>
      <RCTDatePicker
        date={1550919237944}
        mode="date"
        onChange={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "height": 216,
          }
        }
      />
    </View>
    <Text>
      Departure Date
    </Text>
    <View>
      <RCTDatePicker
        date={1550919237944}
        minimumDate={1550919237944}
        mode="date"
        onChange={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "height": 216,
          }
        }
      />
    </View>
    <View
      accessible={true}
      isTVSelectable={true}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
    >
      <Text>
        Submit
      </Text>
    </View>
  </View>
</View>
`;

exports[`AddReservationScreen rendering should render the child element 1`] = `
<View
  style={
    Object {
      "flex": 1,
    }
  }
>
  <RCTSafeAreaView
    emulateUnlessSupported={true}
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "white",
          "borderBottomColor": "grey",
          "borderBottomWidth": 1,
          "flexDirection": "row",
          "justifyContent": "space-between",
          "minHeight": 64,
          "paddingHorizontal": 8,
        }
      }
    >
      <View
        accessible={true}
        isTVSelectable={true}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "height": 24,
            "opacity": 1,
            "transform": Array [
              Object {
                "rotate": "90deg",
              },
            ],
            "width": 24,
          }
        }
        testID="#back"
      >
        <Image
          source={
            Object {
              "testUri": "../../../src/static/icons/iconBlackChevronRight.png",
            }
          }
          style={
            Object {
              "height": 24,
              "transform": Array [
                Object {
                  "rotate": "90deg",
                },
              ],
              "width": 24,
            }
          }
        />
      </View>
      <View>
        <Text>
          Add New Reservation
        </Text>
      </View>
      <View />
    </View>
  </RCTSafeAreaView>
  <View>
    <TextInput
      allowFontScaling={true}
      onChangeText={[Function]}
      style={
        Object {
          "borderBottomColor": "grey",
          "borderBottomWidth": 1,
          "height": 50,
          "margin": 8,
        }
      }
      testID="#name"
      underlineColorAndroid="transparent"
      value=""
    />
    <TextInput
      allowFontScaling={true}
      onChangeText={[Function]}
      style={
        Object {
          "borderBottomColor": "grey",
          "borderBottomWidth": 1,
          "height": 50,
          "margin": 8,
        }
      }
      testID="#hotelname"
      underlineColorAndroid="transparent"
      value=""
    />
    <Text>
      Arrival Date
    </Text>
    <View>
      <RCTDatePicker
        date={1550919237926}
        mode="date"
        onChange={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "height": 216,
          }
        }
      />
    </View>
    <Text>
      Departure Date
    </Text>
    <View>
      <RCTDatePicker
        date={1550919237926}
        minimumDate={1550919237926}
        mode="date"
        onChange={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "height": 216,
          }
        }
      />
    </View>
    <View
      accessible={true}
      isTVSelectable={true}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
    >
      <Text>
        Submit
      </Text>
    </View>
  </View>
</View>
`;
